File: android/app/src/main/kotlin/com/demo/grpcdemo/domain/usecase/WatchPriceUseCase.kt
package com.demo.grpcdemo.domain.usecase

import com.demo.grpcdemo.domain.model.PriceUpdate
import com.demo.grpcdemo.domain.repository.CatalogRepository
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

class WatchPriceUseCase @Inject constructor(private val repository: CatalogRepository) {
    operator fun invoke(productId: String): Flow<PriceUpdate> =
        repository.watchPrice(productId)
}

File: flutter/lib/domain/usecase/watch_price_use_case.dart
// flutter/lib/domain/usecase/watch_price_use_case.dart
import 'package:grpc_demo/domain/model/price_update.dart';
import 'package:grpc_demo/domain/repository/catalog_repository.dart';

class WatchPriceUseCase {
  const WatchPriceUseCase(this._repository);
  final CatalogRepository _repository;

  Stream<PriceUpdate> call(String productId) =>
      _repository.watchPrice(productId);
}
